lib/core/unittest/CProgramCountersTest.cc

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+#include "CProgramCountersTest.h"
+
+#include <core/CLogger.h>
+#include <core/CProgramCounters.h>
+#include <core/CRapidXmlParser.h>
+#include <core/CRapidXmlStatePersistInserter.h>
+#include <core/CRapidXmlStateRestoreTraverser.h>
+#include <core/CRegex.h>
+#include <core/CSleep.h>
+#include <core/CThread.h>
+
+#include <stdint.h>
+
+namespace {
+
+const int TEST_COUNTER = 0u;
+
+class CProgramCountersTestRunner : public ml::core::CThread {
+public:
+    CProgramCountersTestRunner() : m_I(0), m_N(0) {}
+
+    void initialise(int i, int n) {
+        m_N = n;
+        m_I = i;
+    }
+
+private:
+    virtual void run() {
+        if (m_I < 6) {
+            ml::core::CProgramCounters::counter(TEST_COUNTER + m_I)++;
+        } else {
+            ml::core::CProgramCounters::counter(TEST_COUNTER + m_I - m_N)--;
+        }
+    }
+
+    virtual void shutdown() {}
+
+    int m_I;
+    int m_N;
+};
+
+} // namespace
+
+CppUnit::Test* CProgramCountersTest::suite() {
```

Comment:
would be good to test some malicious activity[*], counters that do not exist etc, depends of course on removing the abort. 

[*] future versions, mixed version clusters, ... there might be valid reasons counters disappear or show up